solidity/contracts/TokenStaking.sol

```diff
@@ -107,44 +107,166 @@ contract TokenStaking is Authorizations, StakeDelegatable {
+        } else {
+            // If there is an existing delegation, top-up of the stake is
+            // initiated.
+            initiateTopUp(_from, _value, operator, _extraData);
+        }
+    }
+
+    /// @notice Delegates tokens to a new operator using beneficiary and
+    /// authorizer passed in _extraData parameter.
+    /// @param _from The owner of the tokens who approved them to transfer.
+    /// @param _value Approved amount for the transfer and stake.
+    /// @param _operator The new operator address.
+    /// @param _extraData Data for stake delegation as passed to receiveApproval.
+    function delegate(
+        address _from,
+        uint256 _value,
+        address _operator,
+        bytes memory _extraData
+    ) internal {
+        address payable beneficiary = address(uint160(_extraData.toAddress(0)));
+        address authorizer = _extraData.toAddress(40);
+
+        operators[_operator] = Operator(
             OperatorParams.pack(_value, block.timestamp, 0),
             _from,
             beneficiary,
             authorizer
         );
-        ownerOperators[_from].push(operator);
+        ownerOperators[_from].push(_operator);
+
+        grantStaking.tryCapturingDelegationData(
+            tokenGrant,
+            address(escrow),
+            _from,
+            _operator,
+            _extraData
+        );
+
+        emit Staked(_from, _operator, beneficiary, authorizer, _value);
+    }
 
-        if (_from == address(escrow)) {
-            grantStaking.setGrantForOperator(operator, _extraData.toUint(60));
+    /// @notice Initializes top-up to an existing operator. Tokens added in
+    /// a top-up are not included in the operator stake until the initialization
+    /// period for a top-up passes and top-up is committed. Operator must not
+    /// have the stake undelegated and it has to be initialized. It is expected
+    /// that the top-up is done from the same source of tokens as the initial
+    /// delegation. That is, if the tokens were delegated from a grant, top-up
+    /// has to be performed from the same grant. If the delegation was done
+    /// using liquid tokens, only liquid tokens from the same owner can be used
+    /// to top-up the stake.
+    /// @param _from The owner of the tokens who approved them to transfer.
+    /// @param _value Approved amount for the transfer and top-up to
+    /// an existing stake.
+    /// @param _operator The new operator address.
+    /// @param _extraData Data for stake delegation as passed to receiveApproval
+    function initiateTopUp(
+        address _from,
+        uint256 _value,
+        address _operator,
+        bytes memory _extraData
+    ) internal {
+        uint256 operatorParams = operators[_operator].packedParams;
+        require(
+            _isInitialized(operatorParams),
+            "Initialization period is not over"
+        );
+        require(
+            !_isUndelegating(operatorParams),
```

Comment:
It could also be saying that the stake is currently undelegating and it hasn't been recovered yet - the delegation still exists and the operator has to be active. Really depends on the context but I do agree it could be confusing, especially in this line.